markets/bfp-market/contracts/modules/PerpAccountModule.sol

```diff
@@ -118,4 +136,128 @@ contract PerpAccountModule is IPerpAccountModule {
+        PerpMarketConfiguration.Data storage marketConfig,
+        uint256 oraclePrice
+    ) private {
+        // Determine if toPosition can be immediately liquidated.
+        if (Position.isLiquidatable(position, market, oraclePrice, marketConfig, marginValues)) {
+            revert ErrorUtil.CanLiquidatePosition();
+        }
+
+        accountMargin.updateAccountDebtAndCollateral(
+            market,
+            marginValues.marginUsd.toInt() - marginValues.collateralUsd.toInt()
+        );
+    }
+
+    /**
+     * @inheritdoc IPerpAccountModule
+     */
+    function mergeAccounts(uint128 fromId, uint128 toId, uint128 marketId) external {
+        // Check msg sender is owner for both the accounts.
+        Account.loadAccountAndValidatePermission(
+            fromId,
+            AccountRBAC._PERPS_MODIFY_COLLATERAL_PERMISSION // TODO maybe new permission?
+        );
+        Account.loadAccountAndValidatePermission(
+            toId,
+            AccountRBAC._PERPS_MODIFY_COLLATERAL_PERMISSION // TODO maybe new permission?
+        );
+        PerpMarket.Data storage market = PerpMarket.exists(marketId);
+        PerpMarketConfiguration.Data storage marketConfig = PerpMarketConfiguration.load(marketId);
+        Margin.GlobalData storage globalMarginConfig = Margin.load();
+        Margin.Data storage fromAccountMargin = Margin.load(fromId, marketId);
+
+        Runtime_mergeAccounts memory runtime;
+
+        runtime.supportedSynthMarketIdsLength = globalMarginConfig.supportedSynthMarketIds.length;
+
+        for (uint256 i = 0; i < runtime.supportedSynthMarketIdsLength; ) {
+            runtime.synthMarketIdForLoop = globalMarginConfig.supportedSynthMarketIds[i];
+            runtime.fromAccountCollateralForLoop = fromAccountMargin.collaterals[
+                runtime.synthMarketIdForLoop
+            ];
+            if (runtime.fromAccountCollateralForLoop > 0) {
+                if (
+                    globalMarginConfig.supported[runtime.synthMarketIdForLoop].oracleNodeId !=
+                    marketConfig.oracleNodeId
+                ) {
+                    // If the from account have collateral >0 with different oracle node id then the market, revert.
+                    revert ErrorUtil.OracleNodeMismatch();
+                }
+                // This is the collateral we're intrested in and it has correct oracle node id.
+                runtime.synthMarketId = runtime.synthMarketIdForLoop;
+                runtime.fromAccountCollateral = runtime.fromAccountCollateralForLoop;
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        Margin.Data storage toAccountMargin = Margin.load(toId, marketId);
+        Position.Data storage fromPosition = market.positions[fromId];
+        Position.Data storage toPosition = market.positions[toId];
+
+        if (fromPosition.entryTime != block.timestamp) {
+            revert ErrorUtil.PositionTooOld();
```

Comment:
Need to revisit these error names across the entire market tbh. This is a weird name cause it's generic enough that it now introduces an old position concept but it's specific in that it's only ever used in this single location in merge account. I think a better error is `MergeFailed(string reason)` error and you use that everywhere. Same goes for `OracleMismatch` tbh

Reply:
hmm I agree that its a bit too generic. Not sure if I like having a reason string though as then we loose the nice benefit of seeting all our errors in the error file? Maybe just better naming, eg. `MergeAccountOracleMismatch`, `MergeAccountPositionToOld` ?

Reply:
Yep happy to leave it to you then